File: go/src/RumbleRacing/core/asset/TxtR.go
package asset

import (
	"rumble-reader/chunk/shoc"
	"strconv"
)

type TextEntry struct {
	Index int
	Value string
}

type TxtR struct {
	header      shoc.SHDR
	rawData     []byte
	TextEntries []TextEntry
}

func (t *TxtR) GetType() string {
	return "TxtR"
}

func (t *TxtR) RawData() []byte {
	return t.rawData
}

func (t *TxtR) Header() shoc.SHDR {
	return t.header
}

func ParseTxtR(buf []byte, header shoc.SHDR) (*TxtR, error) {
	resource := TxtR{
		header:  header,
		rawData: buf,
	}
	i := 0

	for i < len(buf) {
		start := i
		for i < len(buf) && buf[i] != 0 {
			i++
		}

		if i == start {
			i++
			continue
		}

		s := string(buf[start:i])

		var numPart, textPart string
		for j, ch := range s {
			if ch == ' ' {
				numPart = s[:j]
				textPart = s[j+1:]
				break
			}
		}

		num, err := strconv.Atoi(numPart)

		if err != nil {
			num = -1
		}

		resource.TextEntries = append(resource.TextEntries, TextEntry{Index: num, Value: textPart})

		i++
	}

	return &resource, nil
}

File: src/RumbleRacing/asset/TxtR.ts
import { SHDR } from "../chunk/shoc/shdr";

export interface TextEntry {
  index: number;
  value: string;
}

export interface TxtR {
  kind: "TxtR";
  header: SHDR;
  textEntries: TextEntry[];
  raw: Uint8Array;
}

export function parseTxtR(buf: Uint8Array, header: SHDR): TxtR {
  const textEntries: TextEntry[] = [];
  let i = 0;
  const decoder = new TextDecoder();

  while (i < buf.length) {
    const start = i;
    while (i < buf.length && buf[i] !== 0) i++;

    if (i === start) {
      i++;
      continue;
    }

    const s = decoder.decode(buf.slice(start, i));

    let num = -1;
    let textPart = "";
    const spaceIdx = s.indexOf(" ");
    if (spaceIdx !== -1) {
      const parsed = parseInt(s.slice(0, spaceIdx), 10);
      num = isNaN(parsed) ? -1 : parsed;
      textPart = s.slice(spaceIdx + 1);
    }

    textEntries.push({ index: num, value: textPart });
    i++;
  }

  return { kind: "TxtR", header, textEntries, raw: buf };
}
